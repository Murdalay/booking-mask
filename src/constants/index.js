import keyMirror from 'fbjs/lib/keyMirror';

/**
 * @namespace Constants
 * @desc App constants
 */

/**
 * @constant {Object} ActionTypes
 * @memberof Constants
 */
export const ActionTypes = keyMirror({
  EXCEPTION: undefined,
  TOGGLE_ROUND_TRIP: undefined,
  BOOKING_TOGGLE_PASSANGERS_DROPDOWN: undefined,
  BOOKING_CHANGE_PASSANGERS_NUMBER: undefined,
  BOOKING_CHANGE_FROM: undefined,
  BOOKING_CHANGE_TO: undefined,
  BOOKING_CHANGE_FLIGHT_DATES: undefined,
  BOOKING_SUBMIT: undefined,
  BOOKING_SUBMIT_SUCCESS: undefined,
  BOOKING_SET_INVALID: undefined,
  BOOKING_UNSET_INVALID: undefined,
  BOOKING_URL_READY: undefined,
  AIROPORTS_LIST_GET: undefined,
  AIROPORTS_LIST_GET_SUCCESS: undefined,
  AIROPORTS_LIST_GET_FAILURE: undefined,
  SHOW_ALERT: undefined,
  HIDE_ALERT: undefined,
});

/**
 * @constant {Object} STATUS
 * @memberof Constants
 */
export const STATUS = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
