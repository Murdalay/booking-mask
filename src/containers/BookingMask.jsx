import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled, { ThemeProvider } from 'styled-components';
import treeChanges from 'tree-changes';
import {
  showAlert,
  toggleRoundTrip,
  changePassengersNumber,
  changeFlightDates,
  unsetInvalid,
  changeFrom,
  changeTo,
  airportsListGet,
  togglePassengersDropdown,
  submit,
} from 'actions/index';
import { STATUS } from 'constants/index';
import theme from 'modules/theme';
import { Box } from 'styled-minimal';

// project components
import FlightDates from 'components/FlightDates';
import PassengersSelector from 'components/PassengersSelector';
import DirectionsToggle from 'components/DirectionsToggle';
import FlightDirection from '../components/FlightDirection';
import StyledContainer from '../components/StyledContainer';
import SubmitButton from '../components/SubmitButton';
import BookingMaskGrid from '../components/BookingMaskGrid';
import Spacer from '../components/Spacer';

const GridItem = styled(Box)`
  display: table;
  text-align: center;
  width: 100%;
`;

const Url = styled.h3`
  padding: 2rem;
  text-align: center;
  width: 100%;
`;

export class BookingMask extends React.PureComponent {
  static propTypes = {
    airportsList: PropTypes.object.isRequired,
    booking: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired,
    isRoundTrip: PropTypes.bool.isRequired,
    theme: PropTypes.object,
  };

  static defaultProps = {
    theme: {},
  };

  componentWillReceiveProps(nextProps) {
    const { dispatch } = this.props;
    const { changedTo } = treeChanges(this.props, nextProps);

    if (changedTo('airportsList.status', STATUS.ERROR)) {
      dispatch(showAlert(nextProps.airportsList.message, { variant: 'danger' }));
    }
    // that allows to dismiss open date / route selectors on click
    if (
      changedTo('booking.invalid', 'from') ||
      changedTo('booking.invalid', 'to') ||
      changedTo('booking.invalid', 'departure')
    ) {
      setTimeout(() => window.addEventListener('click', this.dismissInvalid));
    }
  }

  componentDidMount() {
    const { dispatch } = this.props;
    // preloding airoports list
    dispatch(airportsListGet());
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.dismissInvalid);
  }

  /**
   * Extending project level theme with one passed with a props
   * @memberof BookingMask
   */
  getTheme = () => {
    const { theme: propsTheme } = this.props;

    return Object.assign({}, theme, propsTheme);
  };

  /**
   * Flight departure airport change event handler
   * @memberof BookingMask
   */
  handleDirectionToggle = () => {
    const { dispatch } = this.props;

    dispatch(toggleRoundTrip());
  };

  /**
   * Flight departure airport change event handler
   * @memberof BookingMask
   */
  togglePassengersDropdown = () => {
    const { dispatch } = this.props;

    dispatch(togglePassengersDropdown());
  };

  /**
   * Flight departure airport change event handler
   * @param {Object} data
   * @memberof BookingMask
   */
  handleChangeFrom = data => {
    const { dispatch } = this.props;

    dispatch(changeFrom(data.code));
    this.dismissInvalid();
  };

  /**
   * Flight arrival airport change event handler
   * @param {Object} data
   * @memberof BookingMask
   */
  handleChangeTo = data => {
    const { dispatch } = this.props;

    dispatch(changeTo(data.code));
    this.dismissInvalid();
  };

  /**
   * Flight dates change event handler
   * @param {Object} data
   * @memberof BookingMask
   */
  handleDatesChange = data => {
    const { dispatch } = this.props;

    dispatch(changeFlightDates(data));
    this.dismissInvalid();
  };

  /**
   * Passengers number change event handler
   * @param {Object} passengersCountObject
   * @memberof BookingMask
   */
  handlePassengersChange = passengersCountObject => {
    const { dispatch } = this.props;

    dispatch(changePassengersNumber(passengersCountObject));
  };

  /**
   * Reset component's invalid state
   * @memberof BookingMask
   */
  dismissInvalid = () => {
    const { dispatch } = this.props;

    dispatch(unsetInvalid());
    window.removeEventListener('click', this.dismissInvalid);
  };

  /**
   * Submit button click handler
   * @memberof BookingMask
   */
  onSubmit = () => {
    const { dispatch } = this.props;

    dispatch(submit());
  };

  render() {
    const { booking, airportsList } = this.props;
    const url = booking.url ? <Url data-testid="BookingMaskUrl">{booking.url}</Url> : '';
    const resultTheme = this.getTheme();

    return (
      <ThemeProvider theme={resultTheme}>
        <div key="BookingMask" data-testid="BookingMaskWrapper">
          <StyledContainer>
            <BookingMaskGrid data-testid="BookingMaskGrid">
              <GridItem data-testid="BookingMaskGridItem1">
                <FlightDirection
                  from={booking.from}
                  to={booking.to}
                  airportsList={airportsList}
                  fromInvalid={booking.invalid === 'from'}
                  toInvalid={booking.invalid === 'to'}
                  onChangeFrom={this.handleChangeFrom}
                  onChangeTo={this.handleChangeTo}
                  dismissInvalid={this.dismissInvalid}
                  data-testid="BookingMaskFlightDirection"
                />
              </GridItem>
              <GridItem data-testid="BookingMaskGridItem2">
                <FlightDates
                  flightDates={booking.flightDates}
                  isRoundTrip={booking.isRoundTrip}
                  onDatesChange={this.handleDatesChange}
                  isInvalid={booking.invalid === 'departure'}
                  customArrowIcon={
                    <DirectionsToggle
                      oneWay={booking.isRoundTrip}
                      onToggle={this.handleDirectionToggle}
                      animate={airportsList.status === 'running'}
                      data-testid="BookingMaskDirectionsToggle"
                    />
                  }
                />
              </GridItem>
              <GridItem data-testid="BookingMaskGridItem3">
                <PassengersSelector
                  onChange={this.handlePassengersChange}
                  total={booking.passengers.total}
                  adultsCount={booking.passengers.adults}
                  childrenCount={booking.passengers.children}
                  infantsCount={booking.passengers.infants}
                  toggleDropdown={this.togglePassengersDropdown}
                  dropdownIsOpen={booking.passengersDropdownIsOpen}
                />
              </GridItem>
              <Spacer />
              <Spacer />
              <GridItem data-testid="BookingMaskGridItem4">
                <SubmitButton
                  size="lg"
                  type="submit"
                  data-testid="BookingMaskSubmitButton"
                  onClick={this.onSubmit}
                >
                  Submit
                </SubmitButton>
              </GridItem>
            </BookingMaskGrid>
          </StyledContainer>
          {url}
        </div>
      </ThemeProvider>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return {
    booking: state.booking,
    airportsList: state.airportsList,
    isRoundTrip: state.booking.isRoundTrip,
  };
}

export default connect(mapStateToProps)(BookingMask);
