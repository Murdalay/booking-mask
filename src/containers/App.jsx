import React from 'react';
import { Router } from 'react-router-dom';
import Helmet from 'react-helmet';
import styled, { ThemeProvider } from 'styled-components';
import 'bootstrap/dist/css/bootstrap.css';
import history from 'modules/history';
import theme from 'modules/theme';
import config from 'config';
import BookingMask from 'containers/BookingMask';
import { Container, Heading, Screen, utils } from 'styled-minimal';
import SystemAlerts from 'containers/SystemAlerts';
import GlobalStyles from 'components/GlobalStyles';

const Header = styled.div`
  margin-bottom: ${utils.spacer(3)};
  text-align: center;
`;

const AppWrapper = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 1.4rem;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
`;

export default class App extends React.Component {
  render() {
    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <AppWrapper>
            <Helmet
              defer={false}
              htmlAttributes={{ lang: 'pt-br' }}
              encodeSpecialCharacters={true}
              defaultTitle={config.title}
              titleTemplate={`%s | ${config.name}`}
              titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            />
            <Screen key="PageContent" data-testid="PageContentWrapper">
              <Container verticalPadding>
                <Header>
                  <Heading>Booking mask test assignment</Heading>
                </Header>
                <BookingMask />
              </Container>
            </Screen>
            <SystemAlerts />
            <GlobalStyles />
          </AppWrapper>
        </ThemeProvider>
      </Router>
    );
  }
}
