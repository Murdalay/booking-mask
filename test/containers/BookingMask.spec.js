import React from 'react';
import { STATUS } from 'constants/index';
import { BookingMask } from 'containers/BookingMask';

jest.mock('uuid/v4', () => () => 'ABCDE');

const mockDispatch = jest.fn();
const props = {
  dispatch: mockDispatch,
  booking: {
    flightDates: {},
    passengers: {},
  },
  passengers: {
    total: 1,
    adults: 1,
    children: 0,
    infants: 0,
  },
  flightDates: {
    departure: null,
    return: null,
  },
  theme: {
    palette: {},
    sizes: {},
  },
  airportsList: {
    data: [],
  },
};

function setup(ownProps = props) {
  return shallow(<BookingMask {...ownProps} />, { attachTo: document.getElementById('react') });
}

describe('BookingMask', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a FlightDirection', () => {
    expect(wrapper.find('WithTheme(FlightDirection)')).toExist();
  });

  it('should render a FlightDates', () => {
    expect(wrapper.find('WithTheme(FlightDates)')).toExist();
  });

  it('should render a PassengersSelector', () => {
    expect(wrapper.find('PassengersSelector')).toExist();
  });

  it('should have dispatched an action on mount', () => {
    expect(mockDispatch).toHaveBeenCalledWith({
      payload: {},
      type: 'AIRPORTS_LIST_GET',
    });
  });

  it('should dispatch an alert', () => {
    wrapper.setProps({
      booking: {
        passengers: {},
        flightDates: {},
      },
      airportsList: {
        data: {},
        status: STATUS.ERROR,
        message: 'Nothing found',
      },
    });

    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'SHOW_ALERT',
      payload: {
        id: 'ABCDE',
        icon: undefined,
        message: 'Nothing found',
        position: 'bottom-right',
        variant: 'danger',
        timeout: 0,
      },
    });
  });

  it('should dispatch an action when onSubmit method is called', () => {
    const instance = wrapper.instance();
    instance.onSubmit({ stopPropagation: () => {} });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: {},
      type: 'BOOKING_SUBMIT',
    });
  });

  it('should dispatch an action when dismissInvalid method is called', () => {
    const instance = wrapper.instance();
    instance.dismissInvalid();

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: {},
      type: 'BOOKING_UNSET_INVALID',
    });
  });

  it('should dispatch an action when handleDirectionToggle method is called', () => {
    const instance = wrapper.instance();
    instance.handleDirectionToggle();

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: {},
      type: 'TOGGLE_ROUND_TRIP',
    });
  });
  it('should dispatch an action when togglePassengersDropdown method is called', () => {
    const instance = wrapper.instance();
    instance.togglePassengersDropdown();

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: {},
      type: 'BOOKING_TOGGLE_PASSENGERS_DROPDOWN',
    });
  });
  it('should dispatch an action when handleChangeFrom method is called', () => {
    const instance = wrapper.instance();
    instance.handleChangeFrom({ code: 'FOO' });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { code: 'FOO' },
      type: 'BOOKING_CHANGE_FROM',
    });
  });
  it('should dispatch an action when handleChangeTo method is called', () => {
    const instance = wrapper.instance();
    instance.handleChangeTo({ code: 'FOO' });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { code: 'FOO' },
      type: 'BOOKING_CHANGE_TO',
    });
  });
  it('should dispatch an action when handleDatesChange method is called', () => {
    const instance = wrapper.instance();
    instance.handleDatesChange({ departure: '11-11-11', return: '22-22-22' });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { departure: '11-11-11', return: '22-22-22' },
      type: 'BOOKING_CHANGE_FLIGHT_DATES',
    });
  });
  it('should dispatch an action when handlePassengersChange method is called', () => {
    const instance = wrapper.instance();
    instance.handlePassengersChange({ adults: 1, childs: 1, infants: 1 });

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { data: { adults: 1, childs: 1, infants: 1 } },
      type: 'BOOKING_CHANGE_PASSENGERS_NUMBER',
    });
  });
});
