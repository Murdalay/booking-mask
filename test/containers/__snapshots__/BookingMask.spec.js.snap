// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingMask should render properly 1`] = `
<div
  data-testid="BookingMaskWrapper"
  key="BookingMask"
>
  <React.Memo
    theme={
      Object {
        "breakpoints": Array [
          400,
          768,
          1024,
          1280,
          1920,
        ],
        "button": Object {
          "borderRadius": Object {
            "lg": 0,
            "xl": 0,
            "xs": 0,
          },
          "padding": Object {
            "lg": Array [
              12,
              28,
            ],
            "xl": Array [
              14,
              32,
            ],
          },
        },
        "palette": Object {},
        "sizes": Object {},
      }
    }
  >
    <BookingMaskGrid
      data-testid="BookingMaskGrid"
    >
      <GridItem
        data-testid="BookingMaskGridItem1"
      >
        <FlightDirection
          airoportsList={
            Object {
              "data": Array [],
            }
          }
          data-testid="BookingMaskFlightDirection"
          dismissInvalid={[Function]}
          fromInvalid={false}
          onChangeFrom={[Function]}
          onChangeTo={[Function]}
          theme={
            Object {
              "breakpoints": Array [
                400,
                768,
                1024,
                1280,
                1920,
              ],
              "button": Object {
                "borderRadius": Object {
                  "lg": 0,
                  "xl": 0,
                  "xs": 0,
                },
                "padding": Object {
                  "lg": Array [
                    12,
                    28,
                  ],
                  "xl": Array [
                    14,
                    32,
                  ],
                },
              },
              "palette": Object {},
              "sizes": Object {},
            }
          }
          toInvalid={false}
        />
      </GridItem>
      <GridItem
        data-testid="BookingMaskGridItem2"
      >
        <FlightDates
          customArrowIcon={
            <DirectionsToggle
              animate={false}
              data-testid="BookingMaskDirectionsToggle"
              onToggle={[Function]}
              theme={
                Object {
                  "breakpoints": Array [
                    400,
                    768,
                    1024,
                    1280,
                    1920,
                  ],
                  "button": Object {
                    "borderRadius": Object {
                      "lg": 0,
                      "xl": 0,
                      "xs": 0,
                    },
                    "padding": Object {
                      "lg": Array [
                        12,
                        28,
                      ],
                      "xl": Array [
                        14,
                        32,
                      ],
                    },
                  },
                  "palette": Object {},
                  "sizes": Object {},
                }
              }
            />
          }
          flightDates={Object {}}
          isInvalid={false}
          onDatesChange={[Function]}
          theme={
            Object {
              "breakpoints": Array [
                400,
                768,
                1024,
                1280,
                1920,
              ],
              "button": Object {
                "borderRadius": Object {
                  "lg": 0,
                  "xl": 0,
                  "xs": 0,
                },
                "padding": Object {
                  "lg": Array [
                    12,
                    28,
                  ],
                  "xl": Array [
                    14,
                    32,
                  ],
                },
              },
              "palette": Object {},
              "sizes": Object {},
            }
          }
        />
      </GridItem>
      <GridItem
        data-testid="BookingMaskGridItem3"
      >
        <PassangersSelector
          adultsCount={1}
          adultsDescription="12+ years"
          adultsLabel="Adults"
          adultsMinLimit={1}
          childrenCount={0}
          childrenDescription="2-11 years"
          childrenLabel="Children"
          dropdownIsOpen={true}
          infantsCount={0}
          infantsDescription="0-1 years"
          infantsLabel="Infants"
          maxOfTheKind={20}
          onChange={[Function]}
          theme={
            Object {
              "breakpoints": Array [
                400,
                768,
                1024,
                1280,
                1920,
              ],
              "button": Object {
                "borderRadius": Object {
                  "lg": 0,
                  "xl": 0,
                  "xs": 0,
                },
                "padding": Object {
                  "lg": Array [
                    12,
                    28,
                  ],
                  "xl": Array [
                    14,
                    32,
                  ],
                },
              },
              "palette": Object {},
              "sizes": Object {},
            }
          }
          toggleDropdown={[Function]}
          totalLabel="Passangers"
        />
      </GridItem>
      <Spacer />
      <Spacer />
      <GridItem
        data-testid="BookingMaskGridItem4"
      >
        <React.Memo
          data-testid="BookingMaskSubmitButton"
          onClick={[Function]}
          size="lg"
          theme={
            Object {
              "breakpoints": Array [
                400,
                768,
                1024,
                1280,
                1920,
              ],
              "button": Object {
                "borderRadius": Object {
                  "lg": 0,
                  "xl": 0,
                  "xs": 0,
                },
                "padding": Object {
                  "lg": Array [
                    12,
                    28,
                  ],
                  "xl": Array [
                    14,
                    32,
                  ],
                },
              },
              "palette": Object {},
              "sizes": Object {},
            }
          }
          type="submit"
        >
          Submit
        </React.Memo>
      </GridItem>
    </BookingMaskGrid>
  </React.Memo>
</div>
`;
